cmd/query/app/grpc_handler.go

```diff
@@ -177,3 +180,135 @@ func (g *GRPCHandler) GetDependencies(ctx context.Context, r *api_v2.GetDependen
+	}
+	m, err := g.metricsQueryService.GetLatencies(ctx, &queryParams)
+	if err := g.handleErr("failed to fetch latencies", err); err != nil {
+		return nil, err
+	}
+	return &metrics.GetMetricsResponse{Metrics: *m}, nil
+}
+
+// GetCallRates is the gRPC handler to fetch call rate metrics.
+func (g *GRPCHandler) GetCallRates(ctx context.Context, r *metrics.GetCallRatesRequest) (*metrics.GetMetricsResponse, error) {
+	bqp, err := g.newBaseQueryParameters(r)
+	if err := g.handleErr("failed to build parameters", err); err != nil {
+		return nil, err
+	}
+	queryParams := metricsstore.CallRateQueryParameters{
+		BaseQueryParameters: bqp,
+	}
+	m, err := g.metricsQueryService.GetCallRates(ctx, &queryParams)
+	if err := g.handleErr("failed to fetch call rates", err); err != nil {
+		return nil, err
+	}
+	return &metrics.GetMetricsResponse{Metrics: *m}, nil
+}
+
+// GetErrorRates is the gRPC handler to fetch error rate metrics.
+func (g *GRPCHandler) GetErrorRates(ctx context.Context, r *metrics.GetErrorRatesRequest) (*metrics.GetMetricsResponse, error) {
+	bqp, err := g.newBaseQueryParameters(r)
+	if err := g.handleErr("failed to build parameters", err); err != nil {
+		return nil, err
+	}
+	queryParams := metricsstore.ErrorRateQueryParameters{
+		BaseQueryParameters: bqp,
+	}
+	m, err := g.metricsQueryService.GetErrorRates(ctx, &queryParams)
+	if err := g.handleErr("failed to fetch error rates", err); err != nil {
+		return nil, err
+	}
+	return &metrics.GetMetricsResponse{Metrics: *m}, nil
+}
+
+// GetMinStepDuration is the gRPC handler to fetch the minimum step duration supported by the underlying metrics store.
+func (g *GRPCHandler) GetMinStepDuration(ctx context.Context, _ *metrics.GetMinStepDurationRequest) (*metrics.GetMinStepDurationResponse, error) {
+	minStep, err := g.metricsQueryService.GetMinStepDuration(ctx, &metricsstore.MinStepDurationQueryParameters{})
+	if err := g.handleErr("failed to fetch min step duration", err); err != nil {
+		return nil, err
+	}
+	return &metrics.GetMinStepDurationResponse{MinStep: minStep}, nil
+}
+
+func (g *GRPCHandler) handleErr(msg string, err error) error {
+	if err == nil {
+		return nil
+	}
+	g.logger.Error(msg, zap.Error(err))
+
+	// Avoid wrapping "expected" errors with an "Internal Server" error.
+	if errors.Is(err, disabled.ErrDisabled) {
+		return errGRPCMetricsQueryDisabled
+	}
+	if _, ok := status.FromError(err); ok {
+		return err
+	}
+
+	// Received an "unexpected" error.
+	return status.Errorf(codes.Internal, "%s: %v", msg, err)
+}
+
+func (g *GRPCHandler) newBaseQueryParameters(r interface{}) (bqp metricsstore.BaseQueryParameters, err error) {
+	if r == nil {
+		return bqp, errNilRequest
+	}
+	var baseRequest *metrics.MetricsQueryBaseRequest
+	switch v := r.(type) {
+	case *metrics.GetLatenciesRequest:
+		baseRequest = v.BaseRequest
+	case *metrics.GetCallRatesRequest:
+		baseRequest = v.BaseRequest
+	case *metrics.GetErrorRatesRequest:
+		baseRequest = v.BaseRequest
+	}
```

Comment:
is there a better way?

The reason for taking this approach is to avoid the same nil check in each of the endpoint handlers (`GetLatencies`, etc...) so they can just pass the parent request object in, and have a single place to perform the nil check.
 
The trade-off is given each of these request objects belong to different types, but I need to access the common `BaseRequest` out of each request, a type switch was necessary.